File: PennMobile/src/main/java/com/pennapps/labs/pennmobile/MainActivity.java
package com.pennapps.labs.pennmobile;

import android.Manifest;
import android.content.Context;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Bundle;
import android.os.StrictMode;
import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;
import com.google.android.material.appbar.AppBarLayout;
import com.google.android.material.navigation.NavigationView;
import com.google.android.material.tabs.TabLayout;

import androidx.appcompat.app.ActionBarDrawerToggle;
import androidx.core.app.ActivityCompat;
import androidx.drawerlayout.widget.DrawerLayout;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentStatePagerAdapter;
import androidx.fragment.app.FragmentTransaction;
import androidx.core.content.ContextCompat;
import androidx.preference.PreferenceManager;
import androidx.viewpager.widget.ViewPager;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;

import android.util.Log;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.RelativeLayout;
import android.widget.Toast;
import com.google.firebase.analytics.FirebaseAnalytics;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.reflect.TypeToken;
import com.pennapps.labs.pennmobile.ExpandedBottomNavBar.ExpandableBottomTabBar;
import com.pennapps.labs.pennmobile.api.Labs;
import com.pennapps.labs.pennmobile.api.PennInTouchNetworkManager;
import com.pennapps.labs.pennmobile.api.Platform;
import com.pennapps.labs.pennmobile.api.Serializer;
import com.pennapps.labs.pennmobile.classes.Building;
import com.pennapps.labs.pennmobile.classes.BusRoute;
import com.pennapps.labs.pennmobile.classes.BusStop;
import com.pennapps.labs.pennmobile.classes.Course;
import com.pennapps.labs.pennmobile.classes.DiningHall;
import com.pennapps.labs.pennmobile.classes.FlingEvent;
import com.pennapps.labs.pennmobile.classes.GSRLocation;
import com.pennapps.labs.pennmobile.classes.GSRReservation;
import com.pennapps.labs.pennmobile.classes.Gym;
import com.pennapps.labs.pennmobile.classes.HomeCell;
import com.pennapps.labs.pennmobile.classes.LaundryRoom;
import com.pennapps.labs.pennmobile.classes.LaundryRoomSimple;
import com.pennapps.labs.pennmobile.classes.LaundryUsage;
import com.pennapps.labs.pennmobile.classes.Person;
import com.pennapps.labs.pennmobile.classes.User;
import com.pennapps.labs.pennmobile.classes.Venue;

import java.util.List;

import io.fabric.sdk.android.Fabric;
import retrofit.ResponseCallback;
import retrofit.RestAdapter;
import retrofit.RetrofitError;
import retrofit.client.Response;
import retrofit.converter.GsonConverter;

public class MainActivity extends AppCompatActivity {

    private static Labs mLabs;
    private static Platform mPlatform;
    private static final int CODE_MAP = 1;
    private ExpandableBottomTabBar tabBarView;
    private boolean tab_showed;
    private boolean loggedIn;
    private SharedPreferences mSharedPrefs;

    @Override
    protected void onCreate(Bundle savedInstanceState) {

        setTheme(R.style.AppTheme);

        if (Build.VERSION.SDK_INT > 28){
            setTheme(R.style.DarkModeApi29);
        }
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_main);
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        mSharedPrefs = PreferenceManager.getDefaultSharedPreferences(this);

        StrictMode.ThreadPolicy policy = new StrictMode.ThreadPolicy.Builder().permitAll().build();
        StrictMode.setThreadPolicy(policy);

        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(false);
            getSupportActionBar().setHomeButtonEnabled(false);
        }

        // Set default fragment to HomeFragment
        FragmentTransaction tx = getSupportFragmentManager().beginTransaction();
        tx.replace(R.id.content_frame, new HomeFragment());
        tx.commit();

        // Expandable bottom nav bar
        tabBarView = findViewById(R.id.bottom_navigation);
        onExpandableBottomNavigationItemSelected();

        // TODO: testing account methods, put in the right place
        mPlatform = getPlatformInstance();
        mPlatform.getAccessToken("test_auth_code", "",
                "https://pennlabs.org/pennmobile/android/callback/", "",
                new ResponseCallback() {
            @Override
            public void success(Response response) {
                Log.d("Accounts", "access token: " + response.getBody());
            }

            @Override
            public void failure(RetrofitError error) {
                Log.d("Accounts", "Error fetching access token " + error);
            }
        });

        mPlatform.getUser("test_access_token", new ResponseCallback() {
            @Override
            public void success(Response response) {
                Log.d("Accounts", "user: " + response.getBody());
            }

            @Override
            public void failure(RetrofitError error) {
                Log.d("Accounts", "Error getting user " + error);
            }
        });

    }

    private void onExpandableBottomNavigationItemSelected() {
        tabBarView.setOnTabClickedListener(new ExpandableBottomTabBar.OnTabClickedListener() {
            @Override
            public void onTabClicked(View view, int tabPos) {
                Fragment fragment = null;
                switch (tabPos) {
                    case 0:
                        if (MainActivity.this.getSupportFragmentManager().getBackStackEntryCount() > 0) {
                            fragment = new HomeFragment();
                        }
                        break;
                    case 1:
                        fragment = new GsrTabbedFragment();
                        break;
                    case 2:
                        fragment = new DiningFragment();
                        break;
                    case 3:
                        fragment = new LaundryFragment();
                        break;
                    case 5:
                        fragment = new FitnessFragment();
                        break;
                    case 6:
                        fragment = new RegistrarFragment();
                        break;
                    case 7:
                        fragment = new DirectoryFragment();
                        break;
                    case 8:
                        fragment = new NewsFragment();
                        break;
                    case 9:
                        fragment = new FlingFragment();
                        break;
                    case 10:
                        fragment = new SupportFragment();
                        break;
                    case 11:
                        fragment = new SettingsFragment();
                        break;
                    case 12:
                        fragment = new AboutFragment();
                        break;
                }

                MainActivity.this.fragmentTransact(fragment);
            }
        });
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
    }

    @Override
    public void onResume() {
        super.onResume();
        setTitle(R.string.main_title);
    }

    @RequiresApi(api = Build.VERSION_CODES.JELLY_BEAN_MR1)
    public void setSelectedTab(int index) {
        tabBarView.resetFocusOnAllTabs();
        tabBarView.setSelectedTab(index);
    }

    public void closeKeyboard() {
        InputMethodManager inputMethodManager = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
        View view = getCurrentFocus();
        if (view != null) {
            inputMethodManager.hideSoftInputFromWindow(view.getWindowToken(), InputMethodManager.HIDE_NOT_ALWAYS);
        }
    }

    public static Platform getPlatformInstance() {
        if (mPlatform == null) {
            GsonBuilder gsonBuilder = new GsonBuilder();
            Gson gson = gsonBuilder.create();
            RestAdapter restAdapter = new RestAdapter.Builder()
                    .setConverter(new GsonConverter(gson))
                    .setEndpoint("https://platform.pennlabs.org")
                    .build();
            mPlatform = restAdapter.create(Platform.class);
        }
        return mPlatform;
    }

    public static Labs getLabsInstance() {
        if (mLabs == null) {
            GsonBuilder gsonBuilder = new GsonBuilder();
            gsonBuilder.registerTypeAdapter(new TypeToken<List<Course>>() {
            }.getType(), new Serializer.CourseSerializer());
            gsonBuilder.registerTypeAdapter(new TypeToken<List<Building>>() {
            }.getType(), new Serializer.BuildingSerializer());
            gsonBuilder.registerTypeAdapter(new TypeToken<List<Person>>() {
            }.getType(), new Serializer.DataSerializer());
            gsonBuilder.registerTypeAdapter(new TypeToken<List<Venue>>() {
            }.getType(), new Serializer.VenueSerializer());
            gsonBuilder.registerTypeAdapter(new TypeToken<List<BusStop>>() {
            }.getType(), new Serializer.BusStopSerializer());
            gsonBuilder.registerTypeAdapter(DiningHall.class, new Serializer.MenuSerializer());
            gsonBuilder.registerTypeAdapter(BusRoute.class, new Serializer.BusRouteSerializer());
            gsonBuilder.registerTypeAdapter(new TypeToken<List<BusRoute>>() {
            }.getType(), new Serializer.BusRouteListSerializer());
            // gets room
            gsonBuilder.registerTypeAdapter(new TypeToken<LaundryRoom>() {
            }.getType(), new Serializer.LaundryRoomSerializer());
            // gets laundry room list
            gsonBuilder.registerTypeAdapter(new TypeToken<List<LaundryRoomSimple>>() {
            }.getType(), new Serializer.LaundryRoomListSerializer());
            gsonBuilder.registerTypeAdapter(new TypeToken<List<GSRLocation>>() {
            }.getType(), new Serializer.GsrLocationSerializer());
            // gets laundry usage
            gsonBuilder.registerTypeAdapter(new TypeToken<LaundryUsage>() {
            }.getType(), new Serializer.LaundryUsageSerializer());
            // gets laundry preferences (used only for testing)
            gsonBuilder.registerTypeAdapter(new TypeToken<List<Integer>>(){
            }.getType(), new Serializer.LaundryPrefSerializer());
            gsonBuilder.registerTypeAdapter(new TypeToken<List<FlingEvent>>(){
            }.getType(), new Serializer.FlingEventSerializer());
            // gets fitness
            gsonBuilder.registerTypeAdapter(new TypeToken<List<Gym>>(){
            }.getType(), new Serializer.GymSerializer());
            // gets gsr reservations
            gsonBuilder.registerTypeAdapter(new TypeToken<List<GSRReservation>>(){
            }.getType(), new Serializer.GsrReservationSerializer());
            // gets homepage
            gsonBuilder.registerTypeAdapter(new TypeToken<List<HomeCell>>(){
            }.getType(), new Serializer.HomePageSerializer());
            //get user
            gsonBuilder.registerTypeAdapter(User.class, new Serializer.UserSerializer());
            Gson gson = gsonBuilder.create();
            RestAdapter restAdapter = new RestAdapter.Builder()
                    .setConverter(new GsonConverter(gson))
                    .setEndpoint("https://api.pennlabs.org")
                    .build();
            mLabs = restAdapter.create(Labs.class);
        }
        return mLabs;
    }

    public void showErrorToast(final int errorMessage) {
        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                Toast.makeText(getApplicationContext(), errorMessage, Toast.LENGTH_SHORT).show();
            }
        });
    }

    public void addTabs(FragmentStatePagerAdapter pageAdapter, final ViewPager pager, boolean scrollable) {
        if (tab_showed) {
            return;
        }
        final AppBarLayout appBar = (AppBarLayout) findViewById(R.id.appbar);

        final TabLayout tabLayout = (TabLayout) getLayoutInflater().inflate(R.layout.tab_layout, null);
        tabLayout.post(new Runnable() {
            @Override
            public void run() {
                tabLayout.setupWithViewPager(pager);
            }
        });
        if (!scrollable) {
            tabLayout.setTabMode(TabLayout.MODE_FIXED);
        }
        appBar.addView(tabLayout, 1);

        pager.setAdapter(pageAdapter);
        tab_showed = true;
    }

    public void removeTabs() {
        tab_showed = false;
        final AppBarLayout appBar = (AppBarLayout) findViewById(R.id.appbar);
        if (appBar != null && appBar.getChildCount() >= 3) {
            appBar.removeViewAt(1);
        }
    }

    public void openMapDirectionMenu() {
        final AppBarLayout appBar = (AppBarLayout) findViewById(R.id.appbar);
        appBar.getChildAt(0).setVisibility(View.GONE);
        appBar.getChildAt(1).setVisibility(View.VISIBLE);
    }

    public void closeMapDirectionMenu() {
        final AppBarLayout appBar = (AppBarLayout) findViewById(R.id.appbar);
        appBar.getChildAt(0).setVisibility(View.VISIBLE);
        appBar.getChildAt(1).setVisibility(View.GONE);
    }

    public RelativeLayout getMenuMapExtension() {
        return (RelativeLayout) findViewById(R.id.menu_map_extension);
    }

    private void getPermission() {
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) !=
                PackageManager.PERMISSION_GRANTED ||
                ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION) !=
                        PackageManager.PERMISSION_GRANTED) {

            ActivityCompat
                    .requestPermissions(this, new String[]{Manifest.permission.ACCESS_FINE_LOCATION,
                            Manifest.permission.ACCESS_COARSE_LOCATION}, CODE_MAP);
        } else {
            Fragment fragment = new MapFragment();
            fragmentTransact(fragment);
        }
    }

    private void fragmentTransact(Fragment fragment) {
        if (fragment != null) {
            final Fragment frag = fragment;
            runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    try {
                        FragmentManager fragmentManager = getSupportFragmentManager();
                        fragmentManager.beginTransaction()
                                .replace(R.id.content_frame, frag)
                                .addToBackStack("Main Activity")
                                .setTransition(FragmentTransaction.TRANSIT_FRAGMENT_OPEN)
                                .commit();
                    } catch (IllegalStateException e) {
                        //ignore because the onSaveInstanceState etc states are called when activity is going to background etc
                    }
                }
            });
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String permissions[], @NonNull int[] grantResults) {
        if (grantResults.length == 0 || grantResults[0] == PackageManager.PERMISSION_DENIED) {
            switch (requestCode) {
                case SaveContactsFragment.permission_read:
                    showErrorToast(R.string.ask_contacts_fail);
                    break;
                default:
                    showErrorToast(R.string.ask_location_fail);
                    break;
            }
            return;
        }
        if (requestCode == CODE_MAP) {
            runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    Fragment fragment = new MapFragment();
                    FragmentManager fragmentManager = getSupportFragmentManager();
                    fragmentManager.beginTransaction()
                            .replace(R.id.content_frame, fragment)
                            .addToBackStack(null)
                            .setTransition(FragmentTransaction.TRANSIT_FRAGMENT_OPEN)
                            .commitAllowingStateLoss();
                }
            });
        }
    }
    
}

File: PennMobile/src/main/java/com/pennapps/labs/pennmobile/MainActivity.kt
package com.pennapps.labs.pennmobile

import StudentLifeRf2
import android.Manifest
import android.content.Context
import android.content.SharedPreferences
import android.content.pm.PackageManager
import android.content.res.Configuration
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import android.os.Build
import android.os.Bundle
import android.os.StrictMode
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.InputMethodManager
import android.webkit.CookieManager
import android.widget.TextView
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AppCompatActivity
import androidx.coordinatorlayout.widget.CoordinatorLayout
import androidx.core.content.ContextCompat
import androidx.core.graphics.ColorUtils
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentStatePagerAdapter
import androidx.fragment.app.FragmentTransaction
import androidx.preference.PreferenceManager
import androidx.viewpager.widget.ViewPager
import com.google.android.material.appbar.AppBarLayout
import com.google.android.material.tabs.TabLayout
import com.google.firebase.analytics.FirebaseAnalytics
import com.google.gson.GsonBuilder
import com.google.gson.reflect.TypeToken
import com.pennapps.labs.pennmobile.api.CampusExpress
import com.pennapps.labs.pennmobile.api.OAuth2NetworkManager
import com.pennapps.labs.pennmobile.api.Platform
import com.pennapps.labs.pennmobile.api.Serializer
import com.pennapps.labs.pennmobile.api.StudentLife
import com.pennapps.labs.pennmobile.api.classes.Account
import com.pennapps.labs.pennmobile.api.fragments.LoginFragment
import com.pennapps.labs.pennmobile.components.sneaker.Sneaker
import com.pennapps.labs.pennmobile.databinding.ActivityMainBinding
import com.pennapps.labs.pennmobile.dining.classes.DiningHall
import com.pennapps.labs.pennmobile.dining.classes.Venue
import com.pennapps.labs.pennmobile.fling.classes.FlingEvent
import com.pennapps.labs.pennmobile.gsr.classes.GSRLocation
import com.pennapps.labs.pennmobile.gsr.classes.GSRReservation
import com.pennapps.labs.pennmobile.home.classes.Post
import com.pennapps.labs.pennmobile.laundry.classes.LaundryRoom
import com.pennapps.labs.pennmobile.more.classes.Contact
import com.pennapps.labs.pennmobile.more.fragments.SaveContactsFragment
import com.pennapps.labs.pennmobile.utils.Utils
import eightbitlab.com.blurview.BlurView
import kotlinx.coroutines.sync.Mutex
import okhttp3.OkHttpClient
import retrofit.RestAdapter
import retrofit.android.AndroidLog
import retrofit.client.OkClient
import retrofit.converter.GsonConverter
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import retrofit2.converter.scalars.ScalarsConverterFactory
import java.util.concurrent.TimeUnit
import com.squareup.okhttp.OkHttpClient as SquareOkHttpClient

class MainActivity : AppCompatActivity() {
    private var tabShowed = false
    private lateinit var fragmentManager: FragmentManager
    private lateinit var mSharedPrefs: SharedPreferences
    private lateinit var binding: ActivityMainBinding

    val tokenMutex = Mutex()
    private lateinit var mFirebaseAnalytics: FirebaseAnalytics
    val mNetworkManager by lazy { OAuth2NetworkManager(this) }

    // Declare the launcher at the top of your Activity/Fragment:
    private val requestPermissionLauncher =
        registerForActivityResult(
            ActivityResultContracts.RequestPermission(),
        ) { isGranted: Boolean ->
            if (isGranted) {
                // FCM SDK (and your app) can post notifications.
            } else {
                // TODO: Inform user that that your app will not show notifications.
            }
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        if (Build.VERSION.SDK_INT > 28) {
            setTheme(R.style.DarkModeApi29)
        }
        super.onCreate(savedInstanceState)
        setTheme(R.style.AppTheme)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)
        if (applicationContext.resources.configuration.uiMode and
            Configuration.UI_MODE_NIGHT_MASK == Configuration.UI_MODE_NIGHT_YES
        ) {
            setTheme(R.style.DarkBackground)
        }
        Utils.getCurrentSystemTime()
        askNotificationPermission()

        setSupportActionBar(binding.include.toolbar)
        fragmentManager = supportFragmentManager
        supportActionBar?.setDisplayHomeAsUpEnabled(false)
        supportActionBar?.setHomeButtonEnabled(false)
        mSharedPrefs = PreferenceManager.getDefaultSharedPreferences(this)

        val policy =
            StrictMode.ThreadPolicy
                .Builder()
                .permitAll()
                .build()
        StrictMode.setThreadPolicy(policy)

        onExpandableBottomNavigationItemSelected()
        showBottomBar()
        supportActionBar?.setDisplayShowTitleEnabled(false)

        mFirebaseAnalytics = FirebaseAnalytics.getInstance(this)
        mFirebaseAnalytics.logEvent("MainActivityStart", null)

        // Show HomeFragment if logged in, otherwise show LoginFragment
        val pennKey = mSharedPrefs.getString(getString(R.string.pennkey), null)
        val guestMode = mSharedPrefs.getBoolean(getString(R.string.guest_mode), false)
        if (pennKey == null && !guestMode) {
            startLoginFragment()
        } else {
            startHomeFragment()
        }

        // Did diningWidgetIntentSetup not as separate function as for some reason when
        // diningWidgetBroadcast out of onCreate setTab does not trigger.
        var diningWidgetBroadCast = 0
        if (intent != null) {
            diningWidgetBroadCast = intent.getIntExtra("Widget_Tab_Switch", -1)
        }
        if (diningWidgetBroadCast != -1) {
            setTab(DINING_ID)
        }
    }

    private fun askNotificationPermission() {
        // This is only necessary for API level >= 33 (TIRAMISU)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            if (ContextCompat.checkSelfPermission(this, Manifest.permission.POST_NOTIFICATIONS) ==
                PackageManager.PERMISSION_GRANTED
            ) {
                // FCM SDK (and your app) can post notifications.
            } else if (shouldShowRequestPermissionRationale(Manifest.permission.POST_NOTIFICATIONS)) {
                // TODO: display an educational UI explaining to the user the features that will be enabled
                //       by them granting the POST_NOTIFICATION permission. This UI should provide the user
                //       "OK" and "No thanks" buttons. If the user selects "OK," directly request the permission.
                //       If the user selects "No thanks," allow the user to continue without notifications.
            } else {
                // Directly ask for the permission
                requestPermissionLauncher.launch(Manifest.permission.POST_NOTIFICATIONS)
            }
        }
    }

    private fun onExpandableBottomNavigationItemSelected() {
        binding.include.expandableBottomBar.setOnNavigationItemSelectedListener { item ->
            val position =
                when (item.itemId) {
                    R.id.nav_home -> MainPagerAdapter.HOME_POSITION
                    R.id.nav_dining -> MainPagerAdapter.DINING_POSITION
                    R.id.nav_gsr -> MainPagerAdapter.GSR_POSITION
                    R.id.nav_laundry -> MainPagerAdapter.LAUNDRY_POSITION
                    R.id.nav_more -> MainPagerAdapter.MORE_POSITION
                    else -> MainPagerAdapter.HOME_POSITION
                }
            binding.include.mainViewPager.setCurrentItem(position, false)
            true
        }
    }

    fun setTab(id: Int) {
        binding.include.expandableBottomBar.selectedItemId = id
    }

    fun setSelectedTab(id: Int) {}

    fun closeKeyboard() {
        val inputMethodManager = getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
        val view = currentFocus
        if (view != null) {
            inputMethodManager.hideSoftInputFromWindow(view.windowToken, InputMethodManager.HIDE_NOT_ALWAYS)
        }
    }

    fun startHomeFragment() {
        for (fragment in supportFragmentManager.fragments) {
            if (fragment != null) {
                fragmentManager.beginTransaction().remove(fragment).commit()
            }
        }
        val mainPagerAdapter = MainPagerAdapter(fragmentManager, lifecycle)
        binding.include.mainViewPager.isSaveEnabled = false
        binding.include.mainViewPager.adapter = mainPagerAdapter
        binding.include.mainViewPager.isUserInputEnabled = false
        binding.include.mainViewPager.offscreenPageLimit = 5
        binding.include.mainViewPager.visibility = View.VISIBLE
        binding.include.expandableBottomBar.visibility = View.VISIBLE
        setTab(HOME_ID)
    }

    fun startLoginFragment() {
        CookieManager.getInstance().removeAllCookie()
        val editor = PreferenceManager.getDefaultSharedPreferences(this).edit()
        editor.remove(getString(R.string.penn_password))
        editor.remove(getString(R.string.penn_user))
        editor.remove(getString(R.string.first_name))
        editor.remove(getString(R.string.last_name))
        editor.remove(getString(R.string.email_address))
        editor.remove(getString(R.string.pennkey))
        editor.remove(getString(R.string.accountID))
        editor.remove(getString(R.string.access_token))
        editor.remove(getString(R.string.guest_mode))
        editor.remove(getString(R.string.initials))
        editor.apply()
        val currentFragment = fragmentManager.findFragmentById(R.id.content_frame)
        val fragment: Fragment = LoginFragment()

        // change the fragment only if we're not already on the login fragment
        if (currentFragment == null || currentFragment::class != fragment::class) {
            fragmentManager.popBackStack(null, FragmentManager.POP_BACK_STACK_INCLUSIVE)
            fragmentManager
                .beginTransaction()
                .replace(R.id.content_frame, fragment)
                .setTransition(FragmentTransaction.TRANSIT_FRAGMENT_OPEN)
                .commit()
            binding.include.mainViewPager.visibility = View.GONE
            binding.include.expandableBottomBar.visibility = View.GONE
        }
    }

    fun showErrorToast(errorMessage: Int) {
        runOnUiThread { Toast.makeText(applicationContext, errorMessage, Toast.LENGTH_SHORT).show() }
    }

    fun addTabs(
        pageAdapter: FragmentStatePagerAdapter?,
        pager: ViewPager,
        scrollable: Boolean,
    ) {
        if (tabShowed) {
            return
        }
        val appBar = findViewById<View>(R.id.appbar) as AppBarLayout
        val tabLayout = layoutInflater.inflate(R.layout.tab_layout, null) as TabLayout
        tabLayout.post { tabLayout.setupWithViewPager(pager) }
        if (!scrollable) {
            tabLayout.tabMode = TabLayout.MODE_FIXED
        }
        appBar.addView(tabLayout, 1)
        pager.adapter = pageAdapter
        tabShowed = true
    }

    fun removeTabs() {
        tabShowed = false
        val appBar = findViewById<View>(R.id.appbar) as AppBarLayout

        if (appBar.childCount >= 2) {
            appBar.removeViewAt(1)
        }
    }

    fun fragmentTransact(
        fragment: Fragment?,
        popBackStack: Boolean,
    ) {
        if (fragment != null) {
            runOnUiThread {
                try {
                    if (popBackStack) {
                        fragmentManager.popBackStack(null, FragmentManager.POP_BACK_STACK_INCLUSIVE)
                    }
                    fragmentManager
                        .beginTransaction()
                        .replace(R.id.content_frame, fragment)
                        .addToBackStack(null)
                        .setTransition(FragmentTransaction.TRANSIT_NONE)
                        .commit()
                } catch (e: IllegalStateException) {
                    // ignore because the onSaveInstanceState etc states are called when activity is going to background etc
                }
            }
        }
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray,
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (grantResults.isEmpty() || grantResults[0] == PackageManager.PERMISSION_DENIED) {
            if (requestCode == SaveContactsFragment.PERMISSION_READ) {
                showErrorToast(R.string.ask_contacts_fail)
            }
            return
        }
    }

    override fun setTitle(title: CharSequence) {
        binding.include.appbar
            .findViewById<View>(R.id.toolbar)
            .findViewById<TextView>(R.id.toolbar_title)
            .text = title
    }

    override fun onBackPressed() {
        super.onBackPressed()
        showBottomBar()
    }

    fun hideBottomBar() {
        binding.include.expandableBottomBar.visibility = View.GONE
        val layoutParams = binding.include.contentFrame.layoutParams as CoordinatorLayout.LayoutParams
        layoutParams.setMargins(0, 0, 0, 0)
        binding.include.contentFrame.layoutParams = layoutParams
    }

    fun showBottomBar() {
        binding.include.expandableBottomBar.visibility = View.VISIBLE
        val layoutParams = binding.include.contentFrame.layoutParams as CoordinatorLayout.LayoutParams
        layoutParams.setMargins(0, 0, 0, Utils.dpToPixel(this, 16f))
        binding.include.contentFrame.layoutParams = layoutParams
    }

    companion object {
        const val HOME = 1
        const val GSR = 2
        const val DINING = 3
        const val LAUNDRY = 4
        const val MORE = 5
        const val PCA = 6

        val HOME_ID = R.id.nav_home
        val GSR_ID = R.id.nav_gsr
        val DINING_ID = R.id.nav_dining

        private var mStudentLife: StudentLife? = null
        private var mStudentLifeRf2: StudentLifeRf2? = null
        private var mPlatform: Platform? = null
        private var mCampusExpress: CampusExpress? = null

        @JvmStatic
        val campusExpressInstance: CampusExpress
            get() {
                if (mCampusExpress == null) {
                    val gsonBuilder = GsonBuilder()
                    val gson = gsonBuilder.create()
                    val restAdapter =
                        RestAdapter
                            .Builder()
                            .setConverter(GsonConverter(gson))
                            .setLogLevel(RestAdapter.LogLevel.FULL)
                            .setLog(AndroidLog("Campus Express"))
                            .setEndpoint(Platform.campusExpressBaseUrl)
                            .build()
                    mCampusExpress = restAdapter.create(CampusExpress::class.java)
                }
                return mCampusExpress!!
            }

        @JvmStatic
        val platformInstance: Platform
            get() {
                if (mPlatform == null) {
                    val gsonBuilder = GsonBuilder()
                    val gson = gsonBuilder.create()
                    val restAdapter =
                        RestAdapter
                            .Builder()
                            .setConverter(GsonConverter(gson))
                            .setLogLevel(RestAdapter.LogLevel.FULL)
                            .setLog(AndroidLog("Platform"))
                            .setEndpoint(Platform.platformBaseUrl)
                            .build()
                    mPlatform = restAdapter.create(Platform::class.java)
                }
                return mPlatform!!
            }

        val studentLifeInstanceRf2: StudentLifeRf2
            get() {
                if (mStudentLifeRf2 == null) {
                    val okHttpClient =
                        OkHttpClient
                            .Builder()
                            .connectTimeout(35, TimeUnit.SECONDS)
                            .readTimeout(35, TimeUnit.SECONDS)
                            .writeTimeout(35, TimeUnit.SECONDS)
                            .build()

                    val retrofit =
                        Retrofit
                            .Builder()
                            .baseUrl("https://pennmobile.org/api/")
                            .client(okHttpClient)
                            .addConverterFactory(ScalarsConverterFactory.create())
                            .addConverterFactory(GsonConverterFactory.create())
                            .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                            .build()
                    mStudentLifeRf2 = retrofit.create(StudentLifeRf2::class.java)
                }
                return mStudentLifeRf2!!
            }

        @JvmStatic
        val studentLifeInstance: StudentLife
            get() {
                if (mStudentLife == null) {
                    val gsonBuilder = GsonBuilder()
                    gsonBuilder.registerTypeAdapter(
                        object : TypeToken<MutableList<Contact?>?>() {}.type,
                        Serializer.DataSerializer<Any?>(),
                    )
                    gsonBuilder.registerTypeAdapter(
                        object : TypeToken<MutableList<Venue?>?>() {}.type,
                        Serializer.VenueSerializer(),
                    )
                    gsonBuilder.registerTypeAdapter(
                        DiningHall::class.java,
                        Serializer.MenuSerializer(),
                    )
                    // gets room
                    gsonBuilder.registerTypeAdapter(
                        object : TypeToken<LaundryRoom?>() {}.type,
                        Serializer.LaundryRoomSerializer(),
                    )
                    gsonBuilder.registerTypeAdapter(
                        object : TypeToken<MutableList<GSRLocation?>?>() {}.type,
                        Serializer.GsrLocationSerializer(),
                    )
                    // gets laundry preferences (used only for testing)
                    gsonBuilder.registerTypeAdapter(
                        object : TypeToken<MutableList<Int?>?>() {}.type,
                        Serializer.LaundryPrefSerializer(),
                    )
                    gsonBuilder.registerTypeAdapter(
                        object : TypeToken<MutableList<FlingEvent?>?>() {}.type,
                        Serializer.FlingEventSerializer(),
                    )
                    // gets gsr reservations
                    gsonBuilder.registerTypeAdapter(
                        object : TypeToken<MutableList<GSRReservation?>?>() {}.type,
                        Serializer.GsrReservationSerializer(),
                    )
                    // gets user
                    gsonBuilder.registerTypeAdapter(
                        Account::class.java,
                        Serializer.UserSerializer(),
                    )
                    // gets posts
                    gsonBuilder.registerTypeAdapter(
                        object : TypeToken<MutableList<Post?>?>() {}.type,
                        Serializer.PostsSerializer(),
                    )
                    val gson = gsonBuilder.create()
                    val okHttpClient = SquareOkHttpClient()
                    okHttpClient.setConnectTimeout(35, TimeUnit.SECONDS) // Connection timeout
                    okHttpClient.setReadTimeout(35, TimeUnit.SECONDS) // Read timeout
                    okHttpClient.setWriteTimeout(35, TimeUnit.SECONDS) // Write timeout
                    val restAdapter =
                        RestAdapter
                            .Builder()
                            .setConverter(GsonConverter(gson))
                            .setClient(OkClient(okHttpClient))
                            .setEndpoint("https://pennmobile.org/api")
                            .build()
                    mStudentLife = restAdapter.create(StudentLife::class.java)
                }
                return mStudentLife!!
            }
    }
}

// checks if internet is connected
fun isOnline(context: Context?): Boolean {
    val connectivityManager =
        context?.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
    val capabilities =
        connectivityManager.getNetworkCapabilities(connectivityManager.activeNetwork)
    if (capabilities != null) {
        when {
            capabilities.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR) -> {
                Log.i("Internet", "NetworkCapabilities.TRANSPORT_CELLULAR")
                return true
            }
            capabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI) -> {
                Log.i("Internet", "NetworkCapabilities.TRANSPORT_WIFI")
                return true
            }
            capabilities.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET) -> {
                Log.i("Internet", "NetworkCapabilities.TRANSPORT_ETHERNET")
                return true
            }
        }
    }
    return false
}

/** Shows an error sneaker given a view group with an optional retry function */
fun ViewGroup.showSneakerToast(
    message: String,
    doOnRetry: (() -> Unit)?,
    sneakerColor: Int,
) {
    val sneaker = Sneaker.with(this)
    val view =
        LayoutInflater
            .from(this.context)
            .inflate(R.layout.custom_sneaker_view, sneaker.getView(), false)

    val blurView: BlurView = view.findViewById(R.id.blurView)

    blurView
        .setupWith(this)
        .setFrameClearDrawable(ColorDrawable(Color.TRANSPARENT))
        .setBlurRadius(10f)
        .setOverlayColor(resources.getColor(sneakerColor))

    val retryBtn = view.findViewById<TextView>(R.id.retryButton)
    doOnRetry ?: run { retryBtn.visibility = View.GONE }
    retryBtn.setOnClickListener { doOnRetry?.invoke() }

    val messageView = view.findViewById<TextView>(R.id.errorMessage)
    if (ColorUtils.calculateLuminance(resources.getColor(sneakerColor)) > 0.4) {
        messageView.setTextColor(Color.BLACK)
        view.findViewById<TextView>(R.id.retryButton).setTextColor(Color.BLACK)
    }
    messageView.text = message

    sneaker.sneakCustom(view).setCornerRadius(12, 16).setMessage(message)
}
